lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -219,20 +327,65 @@ static LogicalResult extractConcatToConcatExtract(ExtractOp op,
-  // Narrow Mux
-  // extract(mux(c,a,b)) -> mux(c,extract(a),extract(b))
-  if (auto mux = dyn_cast_or_null<MuxOp>(op.input().getDefiningOp())) {
-    if (mux->hasOneUse()) {
-      auto newT = rewriter.createOrFold<ExtractOp>(
-          mux.getLoc(), op.getType(), mux.trueValue(), op.lowBit());
-      auto newF = rewriter.createOrFold<ExtractOp>(
-          mux.getLoc(), op.getType(), mux.falseValue(), op.lowBit());
-      rewriter.replaceOpWithNewOp<MuxOp>(op, mux.cond(), newT, newF);
-      return success();
-    }
+// Pattern matches on extract(f(a, b)), and transforms f(extract(a), extract(b))
+// for some known f. This is performed by analyzing all usage sites of the
+// result of f(a, b). When this transformation returns true, it mutates all the
+// usage sites of outerExtractOp to reference the newly created narrowed
+// operation.
+static bool narrowExtractWidth(ExtractOp outerExtractOp,
+                               PatternRewriter rewriter) {
+  auto *innerArg = outerExtractOp.input().getDefiningOp();
+
+  if (!innerArg) {
+    return false;
   }
 
+  return llvm::TypeSwitch<Operation *, bool>(innerArg)
+      // The unreferenced leading bits of Add, Sub and Mul can be stripped of,
+      // but not the trailing bits. The trailing bits is used to compute the
+      // results of arithmetic operations.
+      .Case<AddOp, MulOp>([&](auto innerOp) {
+        return narrowOperationWidth(innerOp, innerOp.inputs(),
+                                    /* narrowTrailingBits= */ false, rewriter);
+      })
+      .Case<SubOp>([&](auto innerOp) {
+        return narrowOperationWidth(innerOp, {innerOp.lhs(), innerOp.rhs()},
+                                    /* narrowTrailingBits= */ false, rewriter);
+      })
+      // Bit-wise operations and muxes can be narrowed more aggressively.
+      // Trailing bits that are not referenced in the use-sits can all be
+      // removed.
+      .Case<AndOp, OrOp, XorOp>([&](auto innerOp) {
+        return narrowOperationWidth(innerOp, innerOp.inputs(),
+                                    /* narrowTrailingBits= */ true, rewriter);
+      })
+      .Case<MuxOp>([&](auto innerOp) {
+        Type type = innerOp.getType();
+
+        assert(type.isa<IntegerType>() &&
+               "extract() requires input to be of type IntegerType!");
+
+        // mux(cond, t, f) -> concat(0, mux(cond, t[n-1:0], f[n-1:0])),
+        // where n is the widest demanded width
+        auto cond = innerOp.cond();
+        auto loc = innerOp.getLoc();
+        auto createMuxOp = [&](ArrayRef<Value> values) -> MuxOp {
+          assert(values.size() == 2 &&
+                 "createMuxOp expects exactly two elements");
+          return rewriter.create<MuxOp>(loc, cond, values[0], values[1]);
+        };
+
+        return narrowOperationWidth(
+            innerOp, {innerOp.trueValue(), innerOp.falseValue()},
+            /* narrowTrailingBits= */ true, rewriter, createMuxOp);
+      })
+
```

Comment:
This patch is written to eagerly transform an expression tree in one canonicalization step.  I think it would be simpler (and just as effective) if you transformed `trunc(op x, y)` into `op(trunc x, trunc y)` and then let recursive canonicalization take care of the `trunc x` and `trunc y` as separate steps.  Did you consider that approach?

Reply:
Did you mean transforming `innerArg` into `trunc(cat(zeros, op(trunc(x), trunc(y)), zeros))`, and `extract(cat(..))` do the rest? The concatenation is added to innerArg to make sure that it retains its width.

I considered it. But I decided against it, since rewriting the usage sites directly was straightforward.

Reply:
ok